### ISSUE TITLE
Dictionary source QUERY gets wrong data if PK is not a FIRST column of dictionary

### ISSUE THREAD

**Issue by bvt123** [ClickHouse/ClickHouse#60753]
If I create a dictionary with a source as Clickhouse query, the dictionary is filled with wrong data when the dictionary's PRIMARY KEY is not the first column of the dictionary definition.

The latest version and 23.8 were affected.

https://fiddle.clickhouse.com/10eb598f-1675-4ff9-b01f-169702c406bb

```
CREATE TABLE source
(
    id1 UInt64,
    id2 UInt64,
    val String
)
ENGINE = MergeTree
ORDER BY (id1,id2);

INSERT INTO source VALUES (1, 1, 'v1'), (1, 2, 'v2'),(1, 3, 'v3'), (1, 4, 'v4') ;

CREATE DICTIONARY d1 (id1 Int64, id2 UInt64, val String )
PRIMARY KEY id2
SOURCE(CLICKHOUSE(query 'select id1,id2,val from source'))
LIFETIME(MIN 0 MAX 10)
LAYOUT(FLAT);

select * from d1; -- wrong
select '---';

CREATE DICTIONARY d2 (id2 Int64, id1 UInt64, val String )
PRIMARY KEY id2
SOURCE(CLICKHOUSE(query 'select id2,id1,val from source'))
LIFETIME(MIN 0 MAX 10)
LAYOUT(FLAT);

select * from d2; -- fine
select '---';

CREATE DICTIONARY d3 (id1 Int64, id2 UInt64, val String )
PRIMARY KEY id2
SOURCE(CLICKHOUSE(table 'source'))
LIFETIME(MIN 0 MAX 10)
LAYOUT(FLAT);

select * from d3; --fine

```

**Comment by UnamedRus** [comment]
Related https://github.com/ClickHouse/ClickHouse/issues/45559

**Comment by jbouska567** [comment]
I encountered a similar issue. The problem is that it affects `dictGet`, which silently returns unexpected results. If the dictionary really needs the key to be in the first position, it should at least throw an error when it’s created.

A simple example:
https://fiddle.clickhouse.com/2edd4eab-5bcf-4608-9afa-b9d20977f165

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**Dictionary source QUERY gets wrong data if PK is not a FIRST column of dictionary**

**ClickGap verified:** reproduced it with the script below on `master`; tested 6 version(s) — 6 affected, 0 clean (see Affects); long-standing (not a recent regression); introducing PR not yet pinned.

### Reproduction
```
CREATE TABLE source (id1 UInt64, id2 UInt64, val String) ENGINE = MergeTree ORDER BY (id1, id2);
INSERT INTO source VALUES (1, 1, 'v1'), (1, 2, 'v2'), (1, 3, 'v3'), (1, 4, 'v4');

CREATE DICTIONARY d1 (id1 Int64, id2 UInt64, val String)
PRIMARY KEY id2
SOURCE(CLICKHOUSE(query 'select id1,id2,val from source'))
LIFETIME(MIN 0 MAX 10)
LAYOUT(FLAT);

SELECT * FROM d1;
-- Expected: 4 rows keyed by id2 (1..4) with val v1..v4.
-- Actual:   a single wrong row `1  4  v4` -- the dictionary uses the first
--           selected column id1 (all = 1) as the key, so every row collapses
--           to key 1 and last-write-wins. Declaring id2 first (d2), or using
--           SOURCE(CLICKHOUSE(table 'source')) (d3), both load correctly.
```

### Bisect
- Reproduces on every release line tested down to `v24.1.8.22` (the oldest with a downloadable binary for this repro). The introducing change predates `v24.1.8.22`, so it can't be pinned from release binaries — this is a **long-standing** issue, not a recent regression. A maintainer with full git history can `git blame` the relevant file/line to find the exact change.

### Affects
- **Reproduces on**: `master, 26.5, 26.4, 26.3, 26.2, 26.1`

### Suggested next step
Manual triage of the `comp-dictionary` path is needed. The reproducer above should be a fast first check.

**Component:** `comp-dictionary`
**Severity:** `P2`

---
_ClickGap · Finding: `h_issue_60753`_




### Triage metadata
- Primary component: `comp-dictionary`
- Component owners (`comp-dictionary`): @alexey-milovidov @nihalzp
- Possible duplicate: #74831 — Both report that dictionary QUERY source mishandles data/errors when the primary key column ordering isn't a prefix of the dictionary columns.
- Possibly fixed by (not yet linked): PR #108053
- Related: PR #66595
